### PR TITLE
Add support for inline image Markdown

### PR DESCRIPTION
**Pull request by kidroca**
This Pull Request introduces enhancements to the ExpensiMark library by implementing the conversion of image markdown syntax to HTML `` tags and vice versa. It covers various cases including handling single images and text containing images. 
Additionally, it addresses the conversion of `` tags back to markdown. 
An issue with image alt text containing markdown not converting to plain text is acknowledged and currently marked as a limitation.

### Fixed Issues
$ https://github.com/Expensify/App/issues/37246

# Tests
1. **Unit/Integration Tests Covering the Change:**
   - The changes are covered by unit tests added to `ExpensiMark-HTML-test.js`, `ExpensiMark-HTMLToText-test.js`, and `ExpensiMark-Markdown-test.js`. These tests validate the conversion of image markdown to HTML tags, the handling of images with different attributes, and the conversion of HTML `` tags back to markdown.
   
2. **Tests Performed to Validate Changes:**
   - Ran all newly added tests and ensured they pass, verifying the correct conversion of markdown to HTML `` tags and vice versa.
   - Manually tested the parsing of markdown containing images in various formats and contexts to ensure the output matches expectations.
   - Checked the handling of invalid image URLs to ensure they remain unchanged, as expected.

# QA
1. **QA Validation Steps:**
   - Utilize NewDot to post comments incorporating Markdown (MD) image syntax along with public images sourced from platforms like Unsplash. Verify that these images are correctly displayed inline within the comments.
   - Ensure that you can seamlessly mix text and images within the same comment, maintaining the integrity of the intended message and layout.
   - Confirm the functionality to edit previously submitted comments, checking that upon editing, the comments are accurately presented with the original Markdown syntax intact.
   - Test the handling of invalid Markdown image syntax, such as using a broken or incorrectly formatted URL (e.g., `image`). These should not be parsed and ought to remain displayed as the raw Markdown text, without conversion or rendering as an image.

2. **Areas to Test for Regressions:**
   - General markdown to HTML conversion functionality, especially focusing on links, autolinks, and other media types to ensure no regressions have been introduced.
   - The handling of alt text in both markdown and HTML conversions, given the new logic around image processing.
   - The interaction of image markdown with other markdown features (e.g., bold, italic) to verify that the integration is seamless and does not introduce unexpected behavior.

### CODE REVIEW

File: __tests__/ExpensiMark-HTML-test.js [hunk 1]
```diff
@@ -1886,3 +1886,43 @@ describe('multi-level blockquote', () => {
...
+    });
+
+    // Currently any markdown used inside the square brackets is converted to html string in the alt attribute
+    // The attributes should only contain plain text, but it doesn't seem possible to convert markdown to plain text
+    // or let the parser know not to convert markdown to html for html attributes
+    xtest('Image with alt text containing markdown', () => {
+        const testString = '![*bold* _italic_ ~strike~](https://example.com/image.png)';
+        const resultString = '<img src="https://example.com/image.png" alt="*bold* _italic_ ~strike~" />';
+        expect(parser.replace(testString)).toBe(resultString);
+    });
```
Comment (kidroca) [inline __tests__/ExpensiMark-HTML-test.js:1910]: The interaction of image Markdown with other Markdown features (e.g., bold, italic) currently exhibits partial functionality issues. Specifically, content intended for the `alt` attribute in images is being incorrectly parsed from Markdown to HTML. For instance, given an input like `*bold*`, the output produced is `<img src="https://example/img.jpg" alt="<strong>bold</strong>" />`.
Although this parsing behavior functions as designed and doesn't seem to cause issues in NewDot, the expected outcome for the `alt` attribute should be either `alt="*bold*"` or a plain text `alt="bold"`, without HTML tags. This highlights a known limitation in the current implementation, where Markdown syntax within image alt text does not retain its original or simplified plain text form.

**Comment by situchan** [comment]
reviewing this as C+

**Comment by puneetlath** [comment]
How's the review going @situchan?

File: __tests__/ExpensiMark-HTML-test.js (hunk 1 above)
Comment (kidroca) [reply __tests__/ExpensiMark-HTML-test.js:1910]: Created a fix for this in a separate PR: https://github.com/Expensify/expensify-common/pull/661

**Review by situchan — COMMENTED** [review]
Looks good so far. Still testing various edge cases. Will update soon

**Comment by puneetlath** [comment]
How's the review going @situchan?

**Comment by situchan** [comment]
will complete today. I am also testing with this app [PR](https://github.com/Expensify/App/pull/37566)

**Comment by situchan** [comment]
I will provide feedback on app PR

File: __tests__/ExpensiMark-HTML-test.js [hunk 2]
```diff
@@ -1886,3 +1886,43 @@ describe('multi-level blockquote', () => {
...
+    test('Text containing images', () => {
+        const testString = 'An image of a banana: ![banana](https://example.com/banana.png) an image of a developer: ![dev](https://example.com/developer.png)';
+        const resultString = 'An image of a banana: <img src="https://example.com/banana.png" alt="banana" /> an image of a developer: <img src="https://example.com/developer.png" alt="dev" />';
+        expect(parser.replace(testString)).toBe(resultString);
+    });
+
+    // Currently any markdown used inside the square brackets is converted to html string in the alt attribute
+    // The attributes should only contain plain text, but it doesn't seem possible to convert markdown to plain text
+    // or let the parser know not to convert markdown to html for html attributes
+    xtest('Image with alt text containing markdown', () => {
```
Comment (situchan) [inline __tests__/ExpensiMark-HTML-test.js:1906]: `xtest` ?

Comment (kidroca) [reply __tests__/ExpensiMark-HTML-test.js:1906]: The `xtest` function in Jest is a way to skip a particular test. When you prefix a test with `xtest` instead of `test`, Jest will ignore this test during the test run. This is useful for temporarily disabling a test that might be failing due to unfinished features or bugs that are not yet resolved, without having to comment out the test code. It allows you to easily re-enable the test in the future by changing `xtest` back to `test`.

File: __tests__/ExpensiMark-HTML-test.js [hunk 3]
```diff
@@ -1886,3 +1886,43 @@ describe('multi-level blockquote', () => {
...
+    test('Image with raw data attributes', () => {
+        const testString = '![test](https://example.com/image.png)';
+        const resultString = '<img src="https://example.com/image.png" alt="test" data-raw-href="https://example.com/image.png" data-link-variant="labeled" />';
+        expect(parser.replace(testString, {shouldKeepRawInput: true})).toBe(resultString);
+    });
+
+    test('Image with invalid url should remain unchanged', () => {
+        const testString = '![test](invalid)';
+        expect(parser.replace(testString)).toBe(testString);
+    });
```
Comment (marcaaron) [inline __tests__/ExpensiMark-HTML-test.js:1927]: Should we add a test for something like:

`banana")`

Tbh not sure if xss like that is even remotely possible based on how the markdown parser works.

**Review by marcaaron — APPROVED** [review]
LGTM

File: __tests__/ExpensiMark-HTML-test.js (hunk 3 above)
Comment (kidroca) [reply __tests__/ExpensiMark-HTML-test.js:1927]: That particular example is converted into an `anchor`, presumably due to the use of brackets. However, I will add a simpler test to ensure that additional attributes are not captured.

File: __tests__/ExpensiMark-HTML-test.js [hunk 4]
```diff
@@ -1879,3 +1879,59 @@ describe('multi-level blockquote', () => {
...
+        // "test" and no other attributes
+        const resultString = '![test](<a href=\"https://example.com/image.png\" target=\"_blank\" rel=\"noreferrer noopener\">https://example.com/image.png</a> &quot;title&quot; class=&quot;image&quot;)';
+        expect(parser.replace(testString)).toBe(resultString);
+    });
+
+    test('Trying to inject additional attributes should not work', () => {
+        const testString = '![test" onerror="alert(\'xss\')](https://example.com/image.png)';
+        const resultString = '<img src=\"https://example.com/image.png\" alt=\"test&quot; onerror=&quot;alert(&#x27;xss&#x27;)\" />';
+        expect(parser.replace(testString)).toBe(resultString);
+    });
```
Comment (kidroca) [inline __tests__/ExpensiMark-HTML-test.js:1936]: Updated the PR to include 2 more tests here, other updates are minor code style/formatting changes applied automatically

**Review by situchan — APPROVED** [review]
Looks good! Checklist will be done in [app PR](https://github.com/Expensify/App/pull/37566) after version bump

**Review by marcaaron — APPROVED** [review]
